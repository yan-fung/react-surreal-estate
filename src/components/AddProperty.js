import axios from "axios";
import React, { useState } from "react";
import "../styles/add-property.css";
import Alert from "./Alert";

const AddProperty = () => {
  const initialState = {
    fields: {
      title: "",
      city: "Manchester",
      type: "",
      bedrooms: 1,
      bathrooms: 1,
      price: 0,
      email: "",
    },
    alert: {
      message: "",
      isSuccess: false,
    },
  };

  const [fields, setFields] = useState(initialState.fields);
  const [alert, setAlert] = useState(initialState.alert);

  const handleAddProperty = (event) => {
    event.preventDefault();
    setAlert({ message: "", isSuccess: false });
    axios
      .post("http://localhost:4000/api/v1/PropertyListing", fields)
      .then((res) => {
        setAlert({
          message: "Property added.",
          isSuccess: true,
        });
        setFields({
          title: "",
          city: "Manchester",
          type: "",
          bedrooms: 1,
          bathrooms: 1,
          price: 0,
          email: "",
        });
        console.log(res.data);
      })
      .catch((error) => {
        setAlert({
          message: "Server error, please try again later.",
          isSuccess: false,
        });
        console.log(error);
      });
  };

  const handleFieldChange = (event) => {
    setFields({ ...fields, [event.target.name]: event.target.value });
  };

  return (
    <div className="add-property">
      <h1>Add Properties</h1>
      <Alert message={alert.message} success={alert.isSuccess} />
      <div className="form-container">
        <form className="form-container-2" onSubmit={handleAddProperty}>
          <div className="label-container">
            <label className="label-header" htmlFor="title">
              Title
            </label>
            <input
              id="title"
              name="title"
              value={fields.title}
              placeholder="two bed flat"
              onChange={handleFieldChange}
            />
          </div>
          <div className="label-container">
            <label className="label-header" htmlFor="city">
              City
            </label>
            <select
              id="city"
              name="city"
              value={fields.city}
              onChange={handleFieldChange}
            >
              <option value="Manchester">Manchester</option>
              <option value="Leeds">Leeds</option>
              <option value="Sheffield">Sheffield</option>
              <option value="Liverpool">Liverpool</option>
            </select>
          </div>
          <div className="label-container">
            <label className="label-header" htmlFor="type">
              Property type
            </label>
            <select
              id="type"
              name="type"
              value={fields.type}
              onChange={handleFieldChange}
            >
              <option value="Detached">Detached</option>
              <option value="Semi-Detached">Semi-Detached</option>
              <option value="Terraced">Terraced</option>
              <option value="End of Terrace">End of Terrace</option>
              <option value="Cottage">Cottage</option>
              <option value="Bungalow">Bungalow</option>
            </select>
          </div>
          <div className="label-container">
            <label className="label-header" htmlFor="bedrooms">
              No. of bedrooms
            </label>
            <select
              id="bedrooms"
              name="bedrooms"
              value={fields.bedrooms}
              onChange={handleFieldChange}
            >
              <option value="1">1</option>
              <option value="2">2</option>
              <option value="3">3</option>
              <option value="4">4</option>
              <option value="5">5</option>
              <option value=">5">{`>`} 5</option>
            </select>
          </div>
          <div className="label-container">
            <label className="label-header" htmlFor="bathrooms">
              No. of bathrooms
            </label>
            <select
              id="bathrooms"
              name="bathrooms"
              value={fields.bathrooms}
              onChange={handleFieldChange}
            >
              <option value="1">1</option>
              <option value="2">2</option>
              <option value="3">3</option>
            </select>
          </div>
          <div className="label-container">
            <label className="label-header" htmlFor="price">
              Price (Â£)
            </label>
            <input
              id="price"
              name="price"
              value={fields.price}
              placeholder="1000"
              onChange={handleFieldChange}
            />
          </div>
          <div className="label-container">
            <label className="label-header" htmlFor="email">
              Email
            </label>
            <input
              id="email"
              name="email"
              value={fields.email}
              placeholder="suet0913@hotmail.com"
              onChange={handleFieldChange}
            />
          </div>
          <button className="submit-button" type="submit">
            Add property
          </button>
        </form>
      </div>
    </div>
  );
};

export default AddProperty;
