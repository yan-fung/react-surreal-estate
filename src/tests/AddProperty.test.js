import React from "react";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import AddProperty from "../components/AddProperty";

describe("Add Property", () => {
  it("renders the form of add property correctly", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <AddProperty />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders `Property type` correctly", () => {
    render(
      <MemoryRouter>
        <AddProperty />
      </MemoryRouter>
    );

    const text = screen.getByText("Property type");

    expect(text).toBeInTheDocument("Property type");
  });
});
