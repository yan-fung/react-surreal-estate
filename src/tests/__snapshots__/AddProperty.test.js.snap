// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Property renders the form of add property correctly 1`] = `
<DocumentFragment>
  <div
    class="add-property"
  >
    <h1>
      Add Properties
    </h1>
    <div
      class="form-container"
    >
      <form
        class="form-container-2"
      >
        <div
          class="label-container"
        >
          <label
            class="label-header"
            for="title"
          >
            Title
          </label>
          <input
            id="title"
            name="title"
            placeholder="two bed flat"
            value=""
          />
        </div>
        <div
          class="label-container"
        >
          <label
            class="label-header"
            for="city"
          >
            City
          </label>
          <select
            id="city"
            name="city"
          >
            <option
              value="Manchester"
            >
              Manchester
            </option>
            <option
              value="Leeds"
            >
              Leeds
            </option>
            <option
              value="Sheffield"
            >
              Sheffield
            </option>
            <option
              value="Liverpool"
            >
              Liverpool
            </option>
          </select>
        </div>
        <div
          class="label-container"
        >
          <label
            class="label-header"
            for="type"
          >
            Property type
          </label>
          <select
            id="type"
            name="type"
          >
            <option
              value="Detached"
            >
              Detached
            </option>
            <option
              value="Semi-Detached"
            >
              Semi-Detached
            </option>
            <option
              value="Terraced"
            >
              Terraced
            </option>
            <option
              value="End of Terrace"
            >
              End of Terrace
            </option>
            <option
              value="Cottage"
            >
              Cottage
            </option>
            <option
              value="Bungalow"
            >
              Bungalow
            </option>
          </select>
        </div>
        <div
          class="label-container"
        >
          <label
            class="label-header"
            for="bedrooms"
          >
            No. of bedrooms
          </label>
          <select
            id="bedrooms"
            name="bedrooms"
          >
            <option
              value="1"
            >
              1
            </option>
            <option
              value="2"
            >
              2
            </option>
            <option
              value="3"
            >
              3
            </option>
            <option
              value="4"
            >
              4
            </option>
            <option
              value="5"
            >
              5
            </option>
            <option
              value=">5"
            >
              &gt; 5
            </option>
          </select>
        </div>
        <div
          class="label-container"
        >
          <label
            class="label-header"
            for="bathrooms"
          >
            No. of bathrooms
          </label>
          <select
            id="bathrooms"
            name="bathrooms"
          >
            <option
              value="1"
            >
              1
            </option>
            <option
              value="2"
            >
              2
            </option>
            <option
              value="3"
            >
              3
            </option>
          </select>
        </div>
        <div
          class="label-container"
        >
          <label
            class="label-header"
            for="price"
          >
            Price (Â£)
          </label>
          <input
            id="price"
            name="price"
            placeholder="1000"
            value="0"
          />
        </div>
        <div
          class="label-container"
        >
          <label
            class="label-header"
            for="email"
          >
            Email
          </label>
          <input
            id="email"
            name="email"
            placeholder="suet0913@hotmail.com"
            value=""
          />
        </div>
        <button
          class="submit-button"
          type="submit"
        >
          Add property
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
